sstables/sstables.cc

```diff
@@ -3164,10 +3157,10 @@ future<> replay_pending_delete_log(fs::path pending_delete_log) {
         sstring all(text.begin(), text.end());
         std::vector<sstring> basenames;
         boost::split(basenames, all, boost::is_any_of("\n"), boost::token_compress_on);
-        auto tocs = boost::copy_range<std::vector<sstring>>(basenames
-                | boost::adaptors::filtered([] (auto&& basename) { return !basename.empty(); })
-                | boost::adaptors::transformed([&sstdir] (auto&& basename) { return sstdir + "/" + basename; }));
-        co_await delete_sstables(tocs);
+        auto tocs = boost::copy_range<std::vector<sstring>>(basenames | boost::adaptors::filtered([] (auto&& basename) { return !basename.empty(); }));
+        co_await parallel_for_each(tocs, [&sstdir] (const sstring& name) {
+            return remove_by_toc_name(sstdir + "/" + name);
+        });
     } catch (...) {
```

Comment:
@xemul we don't account for unlinks in I/O scheduler right?

I wonder if we can have significant impact on latency if we push many unlink requests at once, which can happen with large compactions like reshape. Unlink translates into an disk operation for removing the dentry and freeing inode if refcount reaches 0. I don't expect thousands of concurrent unlinks (can happen even in a single shard) to drain bandwidth of high-end disks, but can read and write latencies be potentially affected?

Seastar can't assume though what needs to prioritized. Maybe an application relies on fast deletions for providing good latency.

For the near-term, we can think about limiting deletion concurrency per shard in Scylla. Thoughts?

Reply:
unlink translates to a small number of read/write, it's tiny in the grand scheme of things. You spend millions of iops to create a giant file, a few more iops to delete it is negligible.

What's not negligible is the disk going crazy to discard the data. Nothing much we can do about it though.